src/Routing.jl

```diff
@@ -204,44 +201,61 @@ function update_state_from_channel!(
             x = dist[1]
             distribution = [
                 x / (2 * t * sqrt(π * D * t)) *
-                exp(-((C * t - x)^2 / (4 * D * t))) for t in 1:t_max
+                exp(-((C * t - x)^2 / (4 * D * t))) for
+                t in 1:min(t_max, length(up_timeseries))
             ]
 
-            # perform convolution
-            streamflow[:, 1] =
-                DSP.conv(up_q[:, 1], distribution)[1:size(up_q)[1]]
+            streamflow = dot(up_q[:], distribution[end:-1:1]) # q(t) = sum(q(s)*dist(t-s))
 
-            up_streamflow += streamflow
+            new_state[basin_id] += streamflow # get final streamflow
         end
 
-        # Get dates array
-        dates = timeseries_df.date
-
-        channel_streamflow_df =
-            DataFrame(date = dates, streamflow = up_streamflow[:, 1])
-
-        # Override streamflow timeseries
-        CSV.write(
-            joinpath(output_dir, "channel_basin_$basin_id.csv"),
-            channel_streamflow_df,
-        )
     end
+    return new_state
+
 end
 
 
-function update_state_from_channel!(
-    river_state::RS,
+function compute_channel_state(
+    new_hillslope::Dict,
+    date_window::DateWindow,
     channel_model::CM,
     env::EE,
-    start_date::Date,
-    end_date::Date,
-) where {RS <: RiverState, CM <: AbstractChannelModel, EE <: Environment}
-    update_state_from_channel!(
-        river_state,
+) where {CM <: AbstractChannelModel, EE <: Environment}
+    compute_channel_state(
+        new_hillslope,
+        date_window,
         channel_model,
         env.static_env,
         env.dynamic_env,
-        start_date,
-        end_date,
     )
 end
+
+function compute_streamflow(
+    river_state::RS,
+    static_env::SE,
+    dynamic_env::DE,
+) where {RS <: RiverState, SE <: StaticEnvironment, DE <: DynamicEnvironment}
+
+    output_dir = dynamic_env.output_dir
+    all_basin_ids = static_env.basin_ids
+
+    hillslope_state = river_state.hillslope_state
+    channel_state = river_state.channel_state
+    date_window = river_state.date_window
+
+    total_streamflow =
+        Dict(eachrow([all_basin_ids repeat([NaN], length(all_basin_ids))]))
+    for basin_id in all_basin_ids
+        total_streamflow[basin_id] =
+            hillslope_state[basin_id][end] + channel_state[basin_id]
+    end
+    return total_streamflow
+end
+
+function compute_streamflow(
```

Comment:
Hmmm there is also the method ```
    streamflows, river_states =
        compute_streamflow(initial_window, river_model, env, data_end_date)
```
This makes sense to me (it can dispatch off of the type of the river model), but the "compute_streamflow" function defined here on line 256 is indep of the river model. 


1. Do you envision this method as being independent of the river model type? (will all river models work with this function?)
2. Should these two method actually be two different functions?